saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/serviceprovider/provider/InMemoryRelyingPartyRegistrationRepository.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2002-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.saml2.serviceprovider.provider;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.function.Function;
+
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+
+import static java.util.Arrays.asList;
+import static org.springframework.util.Assert.notEmpty;
+import static org.springframework.util.Assert.notNull;
+
+/**
+ * @since 5.2
+ */
+public class InMemoryRelyingPartyRegistrationRepository
+		implements RelyingPartyRegistrationRepository, Iterable<RelyingPartyRegistration> {
+
+	private final Map<String, RelyingPartyRegistration> byEntityId;
+	private final Map<String, RelyingPartyRegistration> byRegistrationId;
+	private final RelyingPartyRegistration defaultRegistration;
```

Comment:
I don't think the `defaultRegistration` makes sense being derived from the order of the `RelyingPartyRegistration`. What is the functional purpose of having `defaultRegistration`? 

I think defaulting the registration could lead to confusion for a user. Consider if an id was something like `reallylongidthatishardtoread` and a user is sent a link that includes `realllylongidthatishardtoread`.

Perhaps a better approach is allowing a `RelyingPartyRegistration` with a null registrationId and then if the registration id that is extracted from the URL is null a default would be used.